### PR TITLE
[sparse] Add fast semi-structured spasification kernels

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122350

This PR adds in fast semi-structured sparsification kernels to PyTorch.

These kernels allow for accelerated semi-structured sparsification
kernels in PyTorch.

The kernels have been added as aten native functions

In particular, three new functions have been added:

* `torch._sparse_semi_structured_tile`

This function will return the packed representation and metadata for
both X and X', as well as the thread masks. Note that this applies 2:4
sparsity in a 4x4 tile instead of a 1x4 strip as usual.

* `torch._sparse_semi_structured_apply`

This function takes in an input tensor and thread masks from the above
function and returns a packed representation and metadata from applying
thread masks to the input tensor.

* `torch._sparse_semi_structured_apply_dense`

This function does the same thing as above but instead of returning the
tensor in the sparse representation it returns it in the dense
representation

The subclasses have also been updated to add a new
`prune_dense_static_sort`
classmethod to create sparse tensors with this format. I've added some
additional documentatino on how to calculate the compressed tensors
needed to create a SparseSemiStructuredTensor oneself.

To this end, there are two new helper functions added:
`sparse_semi_structured_tile`
`compute_compressed_swizzled_bitmask`

cc @albanD

Differential Revision: [D56190801](https://our.internmc.facebook.com/intern/diff/D56190801)

### CODE REVIEW

File: aten/src/ATen/native/sparse/cuda/SparseSemiStructured.cpp [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#include <torch/library.h>
+#include <torch/types.h>
+
+TORCH_LIBRARY_FRAGMENT(sparse, m) {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/sparse/cuda/SparseSemiStructured.cpp:4]: I don't think this is something we typically land in aten. We'd probalby want to make these private natiev functions.

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Aside from the use of `TORCH_LIBRARY_IMPL` I think this is fine

**Comment by alexsamardzic** [comment]
Will give it a more detailed look, but just a pair of questions at the moment:  Is this PR also replacing what is in `torch.sparse._semi_structrured_conversions` module? If so, shall we remove this module (and, do we have some benchmarking results vs. compiled version of functions in this module)?

**Comment by jcaip** [comment]
@alexsamardzic 

No, this shouldn't replace what's in `torch.sparse._semi_structrured_conversions`. I think ideally we would want to express this logic in pure python so it could be torch.compiled like in those functions. However, these kernels are slightly different because I think they apply 2:4 sparsity over a 4x4 tile instead of 1x4 strip. 

I didn't have the time to look into this more, so just landing these as is for now and we can revisit as needed. Would definitely be interested if you had ideas here though.

**Comment by alexsamardzic** [comment]
Ok, looked more carefully into it - so this is just about importing this functionality from xformers, right?

I've benchmarked it, for CUTLASS back-end, against the compiled version of existing conversion code (will denote variant with compiled version of existing code "Triton", and the new variant "C++/CUDA")

<details>
 <summary>Benchmarking script</summary>

```Python
#! /usr/bin/env python

import random

import torch
from torch.testing import make_tensor
import torch.utils.benchmark as benchmark

from torch.sparse.semi_structured import (
    SparseSemiStructuredTensor,
    to_sparse_semi_structured,
)

SEMI_STRUCTURED_SUPPORTED_DTYPES = [
    torch.float16,
    torch.bfloat16,
    ###torch.float32,
    ###torch.int8,
]


def rand_sparse_semi_structured(r, c, dtype, device, choice=None):
    pattern = "2by4" if dtype != torch.float32 else "1by2"
    if pattern == "1by2":
        ksparse = 2
        choices = [[0, 1], [1, 0]]
    elif pattern == "2by4":
        ksparse = 4
        choices = [
            [1, 1, 0, 0],
            [1, 0, 1, 0],
            [1, 0, 0, 1],
            [0, 1, 1, 0],
            [0, 1, 0, 1],
            [0, 0, 1, 1],
        ]
    mask_entries = [choice or random.choice(choices) for i in range(r * c // ksparse)]
    mask = torch.tensor(mask_entries, dtype=torch.bool).view(r, c).to(device)
    dense = make_tensor(r, c, dtype=dtype, device=device)
    dense[dense == 0] = 1  # To prevent zeros except where mask applied.
    dense = dense.masked_fill(~mask, 0)
    return dense


def run_benchmark(m, k, device, dtype, results):
    label = f"Triton vs. C++/CUDA ({dtype})"
    sub_label = f"m:{m:5d} | k:{k:5d}"

    x = rand_sparse_semi_structured(m, k, dtype, device)

    to_sparse_semi_structured_compiled = torch.compile(
        lambda x: to_sparse_semi_structured(x)
    )

    SparseSemiStructuredTensor._FORCE_CUTLASS = True
    xsp_1 = to_sparse_semi_structured_compiled(x)
    xsp_2 = SparseSemiStructuredTensor.from_dense_fast(x)

    measurement = benchmark.Timer(
        stmt="to_sparse_semi_structured(x)",
        globals={
            "to_sparse_semi_structured": to_sparse_semi_structured_compiled,
            "x": x,
        },
        label=label,
        sub_label=sub_label,
        description="Triton",
    ).blocked_autorange()
    results.append(measurement)

    measurement = benchmark.Timer(
        stmt="to_sparse_semi_structured(x)",
        globals={
            "to_sparse_semi_structured": SparseSemiStructuredTensor.from_dense_fast,
            "x": x,
        },
        label=label,
        sub_label=sub_label,
        description="C++/CUDA",
    ).blocked_autorange()
    results.append(measurement)


if __name__ == "__main__":
    device = "cuda"

    shapes = [
        # distilbert shapes
        (768, 3072, 768),
        (3072, 768, 3072),
        # jiecao shapes
        (1024, 1536, 2048),
        (1024, 9408, 2048),
        (1024, 3200, 2048),
        (1024, 256, 9472),
        (1024, 10240, 256),
        (1024, 256, 12608),
        (1024, 2560, 1024),
        (1024, 512, 10240),
        (1024, 10240, 512),
        (1024, 2048, 1024),
        (1024, 512, 512),
        (1024, 1024, 1024),
        (1024, 2048, 2048),
        (2048, 1536, 2048),
        (2048, 9408, 2048),
        (2048, 3200, 2048),
        (2048, 256, 9472),
        (2048, 10240, 256),
        (2048, 256, 12608),
        (2048, 2560, 1024),
        (2048, 512, 10240),
        (2048, 10240, 512),
        (2048, 2048, 1024),
        (2048, 512, 512),
        (2048, 1024, 1024),
        (2048, 2048, 2048),
    ]

    torch.set_printoptions(
        precision=3,
        threshold=None,
        edgeitems=4,
        linewidth=460,
        profile=None,
        sci_mode=False,
    )

    for dtype in SEMI_STRUCTURED_SUPPORTED_DTYPES:
        results = []

        for m, k, _ in shapes:
            try:
                print(
                    f"m = {m:5d}, k = {k:5d}, dtype = {dtype} ... ",
                    end="",
                )
                run_benchmark(m, k, device, dtype, results)
                print("ok")
            except Exception:
                print("failed")
                continue

        compare = benchmark.Compare(results)
        compare.colorize(rowwise=True)
        compare.print()
```
</details>

"C++/CUDA" variant is indeed reported as significantly faster than "Triton" variant, still each one has its own caveats:

- "C++/CUDA" variant supports `float16` and `bfloat16`, while "Triton" variant supports `int8` and `float32` too.
- "Triton" variant will reuse code compiled for the first shape encountered.  So the real performance could be actually much better than shown by the script above (but still slower than "C++/CUDA" variant, for the same shape), in case if `to_sparse_semi_structured` re-compiled for each individual shape.  That's something to keep in mind in case this C++/CUDA code indeed ported to to-be-compiled PyTorch code.
- Alike, as mentioned in #115008, when "Triton" variant not compiled it uses excessive amount of GPU memory.  This is because it is pretty much written to be compiled, so in order to avoid graph breaks it uses lots of auxiliary tensors.  When compiled, it all gets fused, so it's OK, but if used as is then all of these auxiliary tensors are materialized, so memory usage is huge.  On the other side, PyTorch has issues (or at least it had it when my conversion code merged) with including compiled code.  So that's also something to keep in mind if porting this new C++/CUDA code to PyTorch.

As far as CUTLASS side of things concerned, my opinion is that it should provide fast conversion (at least from dense to sparse semi-structured) routine itself.   It actually provides [code](https://github.com/NVIDIA/cutlass/blob/main/tools/util/include/cutlass/util/host_reorder.h#L86) for the most complicated part of the conversion, but it's CPU code, and implemented as an utility function (i.e. not directly exposed through CUTLASS API).  Still, most of the functionality needed for this conversion is written as host/device functions, so it should be possible to write corresponding GPU code in a reasonable amount of time.  If we think it would be worthwhile, I can do it at some point, and compare to this xformers code.  On the other side, these issues with compiled code mentioned above would still hold.

File: torch/sparse/semi_structured.py [hunk 2]
```diff
@@ -404,6 +413,24 @@ def from_dense(
             requires_grad=original_tensor.requires_grad,
         )
 
+    @classmethod
+    def from_dense_fast(cls, original_tensor : torch.Tensor, algorithm="") -> "SparseSemiStructuredTensor":
```
Comment (cpuhrsch) [inline torch/sparse/semi_structured.py:417]: Maybe `from_dense_sort` or so? Fast isn't quite enough right, because the shuffle is also in there.

Comment (jcaip) [reply torch/sparse/semi_structured.py:417]: Maybe `from_dense` and `prune_dense` ?

`from_dense` takes a dense matrices with 0s in the right place
`prune_dense`` takes a dense matrix and will put the 0s in the right place as part of the process.

Comment (jcaip) [reply torch/sparse/semi_structured.py]: @cpuhrsch How about `prune_dense_static_sort` or `prune_dense_sorting_network` then?

File: torch/sparse/semi_structured.py [hunk 3]
```diff
@@ -404,6 +410,24 @@ def from_dense(
             requires_grad=original_tensor.requires_grad,
         )
 
+    @classmethod
+    def prune_dense_static_sort(cls, original_tensor : torch.Tensor, algorithm="") -> "SparseSemiStructuredTensor":
```
Comment (cpuhrsch) [inline torch/sparse/semi_structured.py:414]: Would be it useful to add docs that write out the decomposition?

Comment (jcaip) [reply torch/sparse/semi_structured.py:414]: Yeah I can update with the equivalent pytorch ops of what this code is doing and write some docstring.

**Comment by jcaip** [comment]
cc @cpuhrsch @alexsamardzic 

One thing that's been on my mind: Right now the way these ops are written, we have three ops
```
sparse_semi_structured_tile(input) -> (packed, meta, packed_t, meta_t, thread_masks)
sparse_semi_structured_apply(input, thread_masks) -> packed, meta
sparse_semi_structured_apply_dense(input, thread_masks) -> dense_output
```

The static_sort_and_compress is separated from the sparse matmul. So basically what happens is that 


`sparse_semi_structured_tile` is called on the dense input tensor and then:
- if forwards: packed and meta are fed to `cslt_sparse_mm / sparse_semi_structured_linear`, cache everything else
- if backwards: gradient is masked using `thread_masks`, `meta` and `meta_t` are reused. 

We could combine the two together to give an op that has the exact same signature as dense matmul. 

```
_sparse_semi_structured_static_sort_compress_sparse_mm(denseA, denseB, transposeA=False, matmul_backend=cusparselt/cutlass, layout=strip/tile) -> dense output, thread_masks
```

You would still need the autograd function to keep track of the metadata to mask the gradient in the backwards for training, but this would give us the option to do a nice autocast-esque UI for inference.

I also see this making composability with something like distributed much easier. If we prune at runtime, all we have to do is make sure that we swap out our mm op last, and we should be able to work with any other accelerated inference method. We could also ask cuSPARSELt to create this op for us too, depending on how urgently we want it.

File: torch/sparse/semi_structured.py (hunk 2 above)
Comment (cpuhrsch) [reply torch/sparse/semi_structured.py]: Hm, why `network`? The decomposition should also help a lot.

Comment (jcaip) [reply torch/sparse/semi_structured.py]: It's how we do the sort https://en.wikipedia.org/wiki/Sorting_network

File: torch/sparse/semi_structured.py [hunk 4]
```diff
@@ -400,19 +406,71 @@ def from_dense(
...
+
+                                                  -> compute swizzled bitmask        -> compressed_swizzled_bitmask
+
+
+        The equivalent PyTorch code to create the same five outputs from the dense tensor can be found below:
+        ```
+        from torch.sparse import SparseSemiStructuredTensorCUTLASS
+        from torch.sparse._semi_structured_conversions import _sparse_semi_structured_tile, _compute_compressed_swizzled_bitmask
+
+        pruned = _sparse_semi_structured_tile(dense)
```
Comment (cpuhrsch) [inline torch/sparse/semi_structured.py:452]: Is it possible to express `_sparse_semi_structured_tile` in standard PyTorch (albeit inefficiently)?

File: torch/sparse/_semi_structured_conversions.py [hunk 5]
```diff
@@ -273,3 +279,73 @@ def sparse_semi_structured_to_dense_cutlass(sparse, meta_reordered):
         )
 
     return dense.view(m, 2 * k)
+
+
+def _sparse_semi_structured_tile(dense):
```
Comment (jcaip) [inline torch/sparse/_semi_structured_conversions.py:284]: @cpuhrsch this is how you would express it without using private ops

**Review by cpuhrsch — APPROVED** [review]

**Comment by jithunnair-amd** [comment]
> @pytorchbot revert -m "This broke rocm builds, which is visible on PR as well" -c ignoredsignal

Thanks @malfet. Although, I think in this case, the PR authors got a bit misled by the DrCI report which put the ROCm builds under flaky, saying "You can merge normally! (2 Unrelated Failures)".... not sure if there's something that can be improved wrt DrCI reporting here:
image
cc @huydhn @clee2000

**Comment by malfet** [comment]
> Although, I think in this case, the PR authors got a bit misled by the DrCI report which put the ROCm builds under flaky, saying "You can merge normally! (2 Unrelated Failures)".... not sure if there's something that can be improved wrt DrCI reporting here

Sorry, I though I left a comment about the same on the issue, but then realized I did it in chat thread. Short term answer is that one should be able to send a feedback on whether or not DrCI classification is accurate. But in this case, let spend some time an investigate why
https://github.com/pytorch/test-infra/issues/5063

**Comment by DanilBaibak** [comment]
Hi @jcaip! Sorry, I need to revert your PR because it breaks the internal build. Here you can find more information about the issue - [D56057736 ](https://www.internalfb.com/diff/D56057736).

**Comment by jcaip** [comment]
@ghimport

**Comment by jcaip** [comment]
@jcaip has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56190801).

**Comment by DanilBaibak** [comment]
Hi @jcaip! Sorry, I need to revert your PR because it broke trunk. Here you can find more information about the issue - 
https://hud.pytorch.org/pytorch/pytorch/commit/14b2273b0c58b4000e10b2e441341eeafb7dd2f6
